# Python for Digital Humanities

## Unit #3: Input/Output

* Printing Output
* Prompting User for Input
* Formatting Output

<font color=blue>---------------------------------------------------------------</font>

## 5. Input/Output

Programs are not very useful if they cannot communicate their results to the user or if they do not allow users to provide specific inputs.

### 5.1 Printing Output

We have already seen the `print` statement (e.g., ```print("Hello World")```.  We also can include multiple items in the print statement.  For example:  
```
Month = "December"
Day = 2
Year = 2019
print("The date is", Day, Month, Year, ".")
```

<font color=blue>---------------------------------------------------------------</font>



### 5.2 Prompting User for Input

If we want the user to input information by typing it on the keyboard, we can use the `input` command.  The general format of the command is as follows:

```variable = input("Some prompt to tell the user what to input:  ")```

For example:
```
name = input("Please enter your name:  ")
print("Hello", name)
```
Or, for the printing, some people prefer an _f string_ to embed a variable in a print statement:

```
name=input("Please enter your name:  ")
print(f'Hello {name}, how are you doing?')
```

Finally, we could put a placeholder in for our variable and attach the variable in a _format_ function:

```
name = input("Please enter your name:  ")
print("Hello {}, how are you doing?".format(name))
```


<font color=blue>---------------------------------------------------------------</font>



### 5.3 Formatting Output 

We often want (or need) more control over the format of the output, especially when writing the decimal places for numeric values.  


To format data, we must use _*format strings*_.  These are encoded notations which inform the computer how we want to modify the format.  

#### Formatting Floating Point Numbers (numbers with decimal places)
For numeric values, we use **%F.wf**, where 
* F is the total number of spaces reserved for the output (it will be replaced with a number), 
* w is the number of digits to the right of the decimal point (it will be replaced with a number), 
* f is code for "floating point number".

For example, if we want to allow 8 spaces for a number, and include only 3 decimal places, we could type: **{:8.3f}**.


In [1]:
print("The result is {:8.3f}.".format(3.14159))

The result is    3.142.


Notice that there is a big gap between the word `is` and the number `3`.  The number `3.142` only occupies 5 spaces (4 digits and 1 decimal point). Because we told the computer that we wanted 8 spaces for the number, it put the 5 characters to the far right of the given spaces (i.e., right-justified), leaving 3 blank spaces to the left of the number.

This justification is good if we have multiple items that we want to line up according to the decimal place.

In [2]:
plane = 678.59
hotel = 4 * 275.00
food = 4 * 31.17

print("      Total Expenses      ")
print("Airfare: {:10.2f}".format(plane))
print("Lodging: {:10.2f}".format(hotel))
print("Meals:   {:10.2f}".format(food))

      Total Expenses      
Airfare:     678.59
Lodging:    1100.00
Meals:       124.68


If we are printing sentences, instead of tables of information, we can let Python handle the total number of spaces by not giving it a value for F.

In [3]:
print("The hotel, which cost ${:.2f}, was more expensive than the airfare, at ${:.2f}.".format(hotel, plane))

The hotel, which cost $1100.00, was more expensive than the airfare, at $678.59.


<font color=blue>---------------------------------------------------------------</font>

### What about formatting strings or integers?

These are actually easier -- use %s for strings and %d for integers.

You can include an integer between the percent sign and the letter to specify the number of spaces that you want to use.   For example %12s will allow 12 spaces for a string.

Without the integer, Python will decide on the spacing.

In [4]:
myPaper = "The Decline of Human Speech"
year = 2013
print("I published my paper, '{:s}', in {:d}.".format(myPaper, year))


I published my paper, 'The Decline of Human Speech', in 2013.


<font color=blue>---------------------------------------------------------------</font>

## Activity:  Formatting a table

1. In a previous example, we saw the following code:
```
 plane = 678.59
 hotel = 4 * 275.00
 food = 4 * 31.17

 print("      Total Expenses      ")
 print("Airfare: {:10.2f}".format(plane))
 print("Lodging: {:10.2f}".format(hotel))
 print("Meals:   {:10.2f}".format(food))
```

On the last 3 lines, I used spaces after the colons to ensure that the 10 reserved spaces for the numbers aligned correctly.  Edit the formatting so that 8 spaces are reserved for the labels (e.g., Airfare, Lodging, and Meals).

Hint:  Use `{:8s}` for the string formatting.
    
<font color=blue>---------------------------------------------------------------</font>